src/core/ddsi/src/q_xevent.c

```diff
@@ -578,6 +578,99 @@ static void handle_xevk_entityid (struct nn_xpack *xp, struct xevent_nt *ev)
   nn_xpack_addmsg (xp, ev->u.entityid.msg, 0);
 }
 
+#ifdef DDSI_INCLUDE_SECURITY
+static void send_heartbeat_to_all_readers(struct nn_xpack *xp, struct xevent *ev, struct writer *wr, nn_mtime_t tnow)
+{
+  struct nn_xmsg **msg = NULL;
+  struct whc_state whcst;
+  nn_mtime_t t_next;
+  int hbansreq = 0;
+  unsigned count = 0, n = 0, i;
+
+  ddsrt_mutex_lock (&wr->e.lock);
+
+  whc_get_state(wr->whc, &whcst);
+
+  if (!writer_must_have_hb_scheduled (wr, &whcst))
+  {
+    hbansreq = 1; /* just for trace */
+    msg = NULL; /* Need not send it now, and no need to schedule it for the future */
+    t_next.v = T_NEVER;
+  }
+  else if (!writer_hbcontrol_must_send (wr, &whcst, tnow))
+  {
+    hbansreq = 1; /* just for trace */
+    msg = NULL;
+    t_next.v = tnow.v + writer_hbcontrol_intv (wr, &whcst, tnow);
+  }
+  else
+  {
+    struct wr_prd_match *m;
+    ddsrt_avl_iter_t it;
+
+    hbansreq = writer_hbcontrol_ack_required (wr, &whcst, tnow);
+    t_next.v = tnow.v + writer_hbcontrol_intv (wr, &whcst, tnow);
+
+    for (m = ddsrt_avl_iter_first (&wr_readers_treedef, &wr->readers, &it); m; m = ddsrt_avl_iter_next (&it))
+    {
+      if (m->seq < m->lst_seq)
+        count++;
+    }
+
+    if (count > 0)
+    {
+      msg = ddsrt_malloc(count * sizeof(struct nn_xmsg *));
```

Comment:
I don't quite like counting readers and allocating memory here. I think it can be eliminated by iterating over the writers in a different way that doesn't require the writer to remain locked, using ``ddsrt_avl_lookup_succ``.

As unlocking the writer for each matching reader is a significant change, I'd like you to change it only if it still makes sense when you look into it a bit further. For now, this code is limited to when security is used, so it is not too big a deal. (But as this commit basically implements writer-side filtering, it probably won't be long before that becomes available to application as well 😄 .)

Reply:
Not yet addressed this issue. Will see if it can be changed.